File: substrate/competition_deep_research_gap.py
"""Competition deep research gap matrix (pure, advisory).

Operator-supplied competitor technical decisions only — never invents
competitor claims or scrapes. backlog_mutated is always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

DecisionArea = Literal[
    "source_acquisition",
    "citation_grounding",
    "multi_agent_orchestration",
    "budget_controls",
    "html_native_reading",
    "model_routing",
    "evaluation_harness",
    "unattended_swarm",
]

GapStatus = Literal["ahead", "parity", "behind", "unknown"]

VALID_AREAS = frozenset(
    {
        "source_acquisition",
        "citation_grounding",
        "multi_agent_orchestration",
        "budget_controls",
        "html_native_reading",
        "model_routing",
        "evaluation_harness",
        "unattended_swarm",
    }
)

VALID_STATUS = frozenset({"ahead", "parity", "behind", "unknown"})


class CompetitionDeepResearchGapError(ValueError):
    """Fail-closed validation for competition gap matrix."""


@dataclass(frozen=True)
class CompetitorDecision:
    competitor: str
    area: DecisionArea
    decision_summary: str
    antiek_status: GapStatus
    residual: str | None


@dataclass(frozen=True)
class CompetitionGapMatrix:
    decisions: tuple[CompetitorDecision, ...]
    behind_count: int
    unknown_count: int
    parity_count: int
    ahead_count: int
    residuals: tuple[str, ...]
    backlog_mutated: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "decisions": [
                {
                    "competitor": d.competitor,
                    "area": d.area,
                    "decision_summary": d.decision_summary,
                    "antiek_status": d.antiek_status,
                    **({"residual": d.residual} if d.residual is not None else {}),
                }
                for d in self.decisions
            ],
            "behind_count": self.behind_count,
            "unknown_count": self.unknown_count,
            "parity_count": self.parity_count,
            "ahead_count": self.ahead_count,
            "residuals": list(self.residuals),
            "backlog_mutated": False,
            "notes": list(self.notes),
            "authority": "competition_deep_research_gap_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise CompetitionDeepResearchGapError(f"{field} must be a non-empty string")
    return value.strip()


def build_competition_deep_research_gap(
    *,
    decisions: object,
    focus_areas: object | None = None,
) -> CompetitionGapMatrix:
    """Build gap matrix from operator-supplied competitor decision records."""
    if not isinstance(decisions, list):
        raise CompetitionDeepResearchGapError("decisions must be an array")

    notes: list[str] = [
        "backlog_mutated=false — advisory gap matrix only",
        "competitor decisions are caller-supplied only (no invent / no scrape)",
    ]

    focus: set[str] | None = None
    if focus_areas is not None:
        if not isinstance(focus_areas, list):
            raise CompetitionDeepResearchGapError(
                "focus_areas must be an array when set"
            )
        focus = set()
        for i, a in enumerate(focus_areas):
            if a not in VALID_AREAS:
                raise CompetitionDeepResearchGapError(
                    f"focus_areas[{i}] invalid DecisionArea"
                )
            focus.add(a)

    rows: list[CompetitorDecision] = []
    behind_count = 0
    unknown_count = 0
    parity_count = 0
    ahead_count = 0
    residuals: list[str] = []

    for i, d in enumerate(decisions):
        if not isinstance(d, dict):
            raise CompetitionDeepResearchGapError(f"decisions[{i}] must be an object")
        competitor = _require_nonempty(
            d.get("competitor"), field=f"decisions[{i}].competitor"
        )
        area = d.get("area")
        if area not in VALID_AREAS:
            raise CompetitionDeepResearchGapError(
                f"decisions[{i}].area invalid DecisionArea"
            )
        if focus is not None and area not in focus:
            continue
        decision_summary = _require_nonempty(
            d.get("decision_summary"),
            field=f"decisions[{i}].decision_summary",
        )
        status = d.get("antiek_status")
        if status not in VALID_STATUS:
            raise CompetitionDeepResearchGapError(
                f"decisions[{i}].antiek_status must be ahead|parity|behind|unknown"
            )
        residual_raw = d.get("residual")
        residual: str | None
        if residual_raw is None:
            residual = None
        elif isinstance(residual_raw, str) and residual_raw.strip():
            residual = residual_raw.strip()
        else:
            raise CompetitionDeepResearchGapError(
                f"decisions[{i}].residual must be non-empty string when set"
            )

        rows.append(
            CompetitorDecision(
                competitor=competitor,
                area=area,  # type: ignore[arg-type]
                decision_summary=decision_summary,
                antiek_status=status,  # type: ignore[arg-type]
                residual=residual,
            )
        )

        if status == "behind":
            behind_count += 1
            if residual:
                residuals.append(residual)
            else:
                residuals.append(
                    f"[{area}] {competitor}: gap recorded without residual text"
                )
        elif status == "unknown":
            unknown_count += 1
        elif status == "parity":
            parity_count += 1
        else:
            ahead_count += 1

    if len(decisions) == 0:
        notes.append("no decisions supplied — empty matrix (no invent competitors)")
    notes.append(
        f"counts ahead={ahead_count} parity={parity_count} "
        f"behind={behind_count} unknown={unknown_count}"
    )
    notes.append("backlog_mutated=false")

    return CompetitionGapMatrix(
        decisions=tuple(rows),
        behind_count=behind_count,
        unknown_count=unknown_count,
        parity_count=parity_count,
        ahead_count=ahead_count,
        residuals=tuple(residuals),
        backlog_mutated=False,
        notes=tuple(notes),
        authority="competition_deep_research_gap_advisory",
    )


__all__ = [
    "CompetitionDeepResearchGapError",
    "CompetitionGapMatrix",
    "CompetitorDecision",
    "build_competition_deep_research_gap",
]

File: apps/reading/src/api/competitionDeepResearchGap.ts
/**
 * Competition deep research gap matrix (pure client).
 *
 * Operator vision: highest quality deep research product in the world —
 * study technical decisions made by competition, record gaps, set agents up
 * for perfect execution. This pure layer only accepts caller-supplied
 * competitor decision records — never invents competitor claims or scrapes.
 */

export type DecisionArea =
  | "source_acquisition"
  | "citation_grounding"
  | "multi_agent_orchestration"
  | "budget_controls"
  | "html_native_reading"
  | "model_routing"
  | "evaluation_harness"
  | "unattended_swarm";

export type GapStatus = "ahead" | "parity" | "behind" | "unknown";

export interface CompetitorDecision {
  competitor: string;
  area: DecisionArea;
  /** Operator-authored summary of their technical choice. */
  decision_summary: string;
  /** Antiek status relative to that decision. */
  antiek_status: GapStatus;
  /** Optional actionable gap residual for future agents. */
  residual?: string;
}

export interface CompetitionGapMatrixInput {
  decisions: CompetitorDecision[];
  /** Optional focus areas; empty means all. */
  focus_areas?: DecisionArea[];
}

export interface CompetitionGapMatrix {
  decisions: CompetitorDecision[];
  behind_count: number;
  unknown_count: number;
  parity_count: number;
  ahead_count: number;
  /** Residuals for areas where Antiek is behind (caller-supplied only). */
  residuals: string[];
  /** Always false — pure matrix does not write product backlog. */
  backlog_mutated: false;
  notes: string[];
  authority: "competition_deep_research_gap_advisory";
}

const VALID_AREAS = new Set<DecisionArea>([
  "source_acquisition",
  "citation_grounding",
  "multi_agent_orchestration",
  "budget_controls",
  "html_native_reading",
  "model_routing",
  "evaluation_harness",
  "unattended_swarm",
]);

const VALID_STATUS = new Set<GapStatus>([
  "ahead",
  "parity",
  "behind",
  "unknown",
]);

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Build competition gap matrix from operator-supplied decision records.
 * Never invents competitor facts.
 */
export function buildCompetitionDeepResearchGap(
  input: CompetitionGapMatrixInput,
): CompetitionGapMatrix {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (!Array.isArray(input.decisions)) {
    throw new Error("decisions must be an array");
  }

  const notes: string[] = [
    "backlog_mutated=false — advisory gap matrix only",
    "competitor decisions are caller-supplied only (no invent / no scrape)",
  ];

  let focus: Set<DecisionArea> | null = null;
  if (input.focus_areas !== undefined && input.focus_areas !== null) {
    if (!Array.isArray(input.focus_areas)) {
      throw new Error("focus_areas must be an array when set");
    }
    focus = new Set();
    for (let i = 0; i < input.focus_areas.length; i++) {
      const a = input.focus_areas[i];
      if (!VALID_AREAS.has(a)) {
        throw new Error(`focus_areas[${i}] invalid DecisionArea`);
      }
      focus.add(a);
    }
  }

  const decisions: CompetitorDecision[] = [];
  let behind_count = 0;
  let unknown_count = 0;
  let parity_count = 0;
  let ahead_count = 0;
  const residuals: string[] = [];

  for (let i = 0; i < input.decisions.length; i++) {
    const d = input.decisions[i];
    if (!d || typeof d !== "object") {
      throw new Error(`decisions[${i}] must be an object`);
    }
    const competitor = requireNonEmpty(
      d.competitor,
      `decisions[${i}].competitor`,
    );
    if (!VALID_AREAS.has(d.area)) {
      throw new Error(`decisions[${i}].area invalid DecisionArea`);
    }
    if (focus && !focus.has(d.area)) {
      continue;
    }
    const decision_summary = requireNonEmpty(
      d.decision_summary,
      `decisions[${i}].decision_summary`,
    );
    if (!VALID_STATUS.has(d.antiek_status)) {
      throw new Error(
        `decisions[${i}].antiek_status must be ahead|parity|behind|unknown`,
      );
    }
    let residual: string | undefined;
    if (d.residual !== undefined && d.residual !== null) {
      if (typeof d.residual !== "string" || !d.residual.trim()) {
        throw new Error(
          `decisions[${i}].residual must be non-empty string when set`,
        );
      }
      residual = d.residual.trim();
    }

    const row: CompetitorDecision = {
      competitor,
      area: d.area,
      decision_summary,
      antiek_status: d.antiek_status,
      residual,
    };
    decisions.push(row);

    if (d.antiek_status === "behind") {
      behind_count += 1;
      if (residual) residuals.push(residual);
      else {
        residuals.push(
          `[${d.area}] ${competitor}: gap recorded without residual text`,
        );
      }
    } else if (d.antiek_status === "unknown") {
      unknown_count += 1;
    } else if (d.antiek_status === "parity") {
      parity_count += 1;
    } else {
      ahead_count += 1;
    }
  }

  if (input.decisions.length === 0) {
    notes.push("no decisions supplied — empty matrix (no invent competitors)");
  }
  notes.push(
    `counts ahead=${ahead_count} parity=${parity_count} behind=${behind_count} unknown=${unknown_count}`,
  );
  notes.push("backlog_mutated=false");

  return {
    decisions,
    behind_count,
    unknown_count,
    parity_count,
    ahead_count,
    residuals,
    backlog_mutated: false,
    notes,
    authority: "competition_deep_research_gap_advisory",
  };
}

export function formatCompetitionGapSummary(m: CompetitionGapMatrix): string {
  return (
    `behind=${m.behind_count} · unknown=${m.unknown_count} · ` +
    `parity=${m.parity_count} · ahead=${m.ahead_count} · backlog_mutated=false`
  );
}
